### PR TITLE
Implement CoCreateGuid using uuid_generate on Unix

### PR DESCRIPTION
**Pull request by janvorli**
We were generating GUIDs on Unix as random numbers. But that is not
correct since GUIDs have defined structure with bits having specific
meanings. For example, there are four bits that represent the type of
the GUID, which means whether the guid is randomly generated,
name based on SHA1, time based etc.
This change changes the CoCreateGuid to use the uuid_generate function
from the uuid library that should generate well-formed GUIDs.

### CODE REVIEW

**Comment by janvorli** [comment]
@jkotas Can you take a look please?

File: src/pal/src/configure.cmake [hunk 1]
```diff
@@ -858,6 +859,11 @@ int main(int argc, char **argv)
         return 0;
 }" UNWIND_CONTEXT_IS_UCONTEXT_T)
 
+if(NOT HAVE_LIBUUID_H)
+  unset(HAVE_LIBUUID_H CACHE)
+  message(FATAL_ERROR "Cannot find libuuid. Try installing uuid-dev or the appropriate packages for your platform")
+endif()
```
Comment (jkotas) [inline src/pal/src/configure.cmake]: Is libuuid installed by default in common distros?

**Comment by janvorli** [comment]
Ah, we will need to install libuuid to our Linux / FreeBSD / CentOS CI machines.

**Comment by jkotas** [comment]
LGTM

**Comment by jkotas** [comment]
And also update "getting started" instructions...

**Comment by janvorli** [comment]
@jkotas It was installed on my OSX, but not on my Ubuntu and obviously from the CI neither on CentOS and FreeBSD.

**Comment by janvorli** [comment]
It fixes #1602

**Comment by janvorli** [comment]
@ellismg Could you please install uuid-dev package on our CI FreeBSD / CentOS machines?

**Comment by mmitche** [comment]
@janvorli I'll look into it

**Comment by janvorli** [comment]
@mmitche thanks!

**Comment by mmitche** [comment]
@janvorli Installed for centos and SuSE.  FreeBSD does not have an obvious package for this.  Can you find out what it is?

**Comment by mmitche** [comment]
@janvorli Also did you update the build documentation and cmake detection to indicate that this needs to be installed.

**Comment by stephentoub** [comment]
This also contributes to https://github.com/dotnet/corefx/issues/3573.  Before this change, Guid.NewGuid() on my Linux VM is ~200x slower than on my Windows machine.  After this change, it's "only" ~60x slower.  Still a significant difference, but quite a nice improvement.

**Comment by janvorli** [comment]
@mmitche I am trying to figure the FreeBSD out. CMake detection is updated, I have locally updated the build doc for Linux, but want to update the FreeBSD too before sending in another checkin.

**Comment by akoeplinger** [comment]
Looks like freebsd already has a similar function built-in: http://www.freebsd.org/cgi/man.cgi?query=uuid_create

**Comment by janvorli** [comment]
@akoeplinger thanks! That's great. I am just installing FreeBSD here so I'll try to use it instead of the lib uuid for FreeBSD

**Comment by adityamandaleeka** [comment]
LGTM modulo others' comments.

**Comment by janvorli** [comment]
@mmitche It looks like either has CentOS a different uuid lib or the header is in a different place. Could you please find the uuid.h somewhere below /usr/include and see if it contains uuid_generate or uuid_create?

**Comment by janvorli** [comment]
@mmitche I've installed a brand new installation of CentOS7.1 into a VM on my machine and it contained the /usr/include/uuid/uuid.h header like I have on Ubuntu.
So I guess that when you've installed uuid for CentOS, you have installed something else, since from the error log on the CI machines, I can see that it has not found the /usr/include/uuid/uuid.h.
I've dumped all of the installed packages on my CentOS that contain uuid in the name using `rpm -qa | grep uuid` command. I can see that I have these packages installed:
`libuuid-devel-2.32.2-21.el7.x86_64`
`libuuid-2.32.2-21.el7.x86_64`

**Comment by mmitche** [comment]
Here's what got installed: Installed Packages
uuid-devel.x86_64                       1.6.2-26.el7                       @base

**Comment by mmitche** [comment]
Well..hmm....libuuid-devel is different apparently.

**Comment by janvorli** [comment]
@dotnet-bot test this please

**Comment by mmitche** [comment]
Should be good now

**Comment by mmitche** [comment]
(assuming I got it in time)

**Pull request merged**